dashboard/app/controllers/xhr_proxy_controller.rb

```diff
@@ -0,0 +1,71 @@
+# A controller for proxying web requests to 3rd party APIs. This protects
+# applab users from XSS attacks, complies with the Same Origin Policy,
+# and prevents http/https mismatch warnings in IE.
+#
+# Responses are cached for one minute since many 3rd party APIs serve data
+# which changes frequently such as news or sports scores.
+#
+# To reduce the likelihood of abuse, we only proxy content with an allowed
+# whitelist of JSON response types. We will need to monitor usage to detect
+# abuse and potentially add other abuse prevention measures.
+
+require 'set'
+
+class XhrProxyController < ApplicationController
+  include ProxyHelper
+
+  ALLOWED_CONTENT_TYPES = Set.new(%w(
+    application/json
+    text/javascript
+    text/json
+    text/plain
+  ))
+
+  # 'code.org' is included so applab apps can access the tables and properties of other applab apps.
+  ALLOWED_HOSTNAME_SUFFIXES = %w(
+    accuweather.com
+    api.data.gov
+    api.randomuser.me
+    api.zippopotam.us
+    code.org
+    data.cityofchicago.org
+    googleapis.com
+    query.yahooapis.com
+    noaa.gov
+    rhcloud.com
+    wikipedia.org
+  )
+
+  # How long the content is allowed to be cached
+  EXPIRY_TIME = 1.minute
+
+  # Return the proxied api at the URL specified in the 'u' parameter. The 'c' parameter
+  # is an unforgeable token which identifies the app lab app which is generating the request,
+  # and may be used to enforce a per-app rate-limit.
+  def get
+    channel_id = params[:c]
+    url = params[:u]
+
+    begin
+      owner_storage_id, _ = storage_decrypt_channel_id(channel_id)
```

Comment:
Note `storage_decrypt_channel_id` returns the (decrypted) channel_id. Should we be comparing it to `params[:c]` (throwing on disagreement)?